### PR TITLE
Add order rule for IEThingWith

### PR DESCRIPTION
**Pull request by mbj**

### CODE REVIEW

File: src/SourceConstraints.hs [hunk 1]
```diff
@@ -179,6 +180,20 @@ sortedIEs dynFlags ies =
...
     ieClass ie = error $ "Unsupported: " <> gshow ie
 
+sortedIEThingWith :: DynFlags -> IE GhcPs -> Maybe SDoc
+sortedIEThingWith dynFlags (IEThingWith _xIE _wrappedName _ieWildcard ieWith _ieFieldLabels) =
+  if rendered /= expected
+     then pure $ text . message $ intercalate ", " expected
+     else empty
+  where
+    message :: String -> String
+    message example = "Unsorted import/export with list: (" <> example <> ")"
```
Comment (snusnu) [inline src/SourceConstraints.hs]: @mbj I find the "with list" wording a bit confusing. What about just

* Unsorted import/export list: ...

OT (and non-blocking): Is there a way to actually be precise and refer to import **or** export?

Comment (mbj) [reply src/SourceConstraints.hs]: > OT (and non-blocking): Is there a way to actually be precise and refer to import **or** export?

Its the same AST node, and I do not have the context on which side we are (import or export)!

Comment (mbj) [reply src/SourceConstraints.hs]: OT: and source constraints would flag on both (import or export!).

Comment (mbj) [reply src/SourceConstraints.hs]: >     * Unsorted import/export list: ...

Well its not an import export list, its a `with` list, see the data type we match here. I think we should stick to the GHC AST terminology. Its a clear difference if we have an unsorted list, or an unsorted list inside a namesapce (and we do not yet know if we are dealing with record fields or constructors, this is not known at this stage!).

Comment (snusnu) [reply src/SourceConstraints.hs]: What about "Unsorted import/export detected: ..." then? I still think the "with" in there is too confusing "grammar" for a human to interpret, even if it matches HS internals most closely.

File: src/SourceConstraints.hs [hunk 2]
```diff
@@ -179,6 +180,20 @@ sortedIEs dynFlags ies =
...
+
+    rendered = render dynFlags <$> ieWith
+    expected = sort rendered
+
+sortedIEThingWith _ _ = empty
```
Comment (snusnu) [inline src/SourceConstraints.hs]: @mbj Wouldn't we typically still provide (`_` prefixed) names even when they're unused?

Comment (mbj) [reply src/SourceConstraints.hs]: IMO only if we'd otherwise not have named them in an adjacent pattern.

**Pull request merged**